**Importing the necessary libraries**

In [2]:
import pandas as pd
import pymongo

**Connection with Mongodb Atlas**

In [3]:
myclient = pymongo.MongoClient("mongodb+srv://kishore-0810:dbatlas123@cluster0.k3nng.mongodb.net/?retryWrites=true&w=majority")
my_db = myclient["sample_airbnb"]
my_coll = my_db["listingsAndReviews"]

**Airbnb Data for First Document**

In [4]:
my_coll.find_one()

{'_id': '10006546',
 'listing_url': 'https://www.airbnb.com/rooms/10006546',
 'name': 'Ribeira Charming Duplex',
 'summary': 'Fantastic duplex apartment with three bedrooms, located in the historic area of Porto, Ribeira (Cube) - UNESCO World Heritage Site. Centenary building fully rehabilitated, without losing their original character.',
 'space': 'Privileged views of the Douro River and Ribeira square, our apartment offers the perfect conditions to discover the history and the charm of Porto. Apartment comfortable, charming, romantic and cozy in the heart of Ribeira. Within walking distance of all the most emblematic places of the city of Porto. The apartment is fully equipped to host 8 people, with cooker, oven, washing machine, dishwasher, microwave, coffee machine (Nespresso) and kettle. The apartment is located in a very typical area of the city that allows to cross with the most picturesque population of the city, welcoming, genuine and happy people that fills the streets with h

**Dataset Description**

In [5]:
# _id: A unique identifier for each listing in the dataset.
# listing_url: The URL of the listing on the Airbnb website.
# name: The title or name of the listing.
# summary: A brief description of the listing.
# space: Detailed information about the space provided in the listing.
# description: A full description of the listing, including space and summary.
# neighborhood_overview: An overview of the neighborhood where the listing is located.
# notes: Additional notes provided by the host about the listing.
# transit: Information about public transportation options near the listing.
# access: Details about what parts of the property guests can access.
# interaction: Information on how much interaction guests can expect from the host.
# house_rules: The rules set by the host for guests.
# property_type: The type of property, such as an apartment, house, or condominium.
# room_type: The type of room being offered, like an entire home, private room, or shared room.
# bed_type: The type of bed provided, such as a real bed, futon, or sofa bed.
# minimum_nights: The minimum number of nights a guest can book.
# maximum_nights: The maximum number of nights a guest can book.
# cancellation_policy: The cancellation policy for the listing.
# last_scraped: The date when the data was last collected from the Airbnb website.
# calendar_last_scraped: The date when the calendar information was last collected.
# first_review: The date of the first review received by the listing.
# last_review: The date of the most recent review received by the listing.
# accommodates: The number of guests the listing can accommodate.
# bedrooms: The number of bedrooms included in the listing.
# beds: The number of beds available in the listing.
# number_of_reviews: The total number of reviews the listing has received.
# bathrooms: The number of bathrooms in the listing.
# amenities: A list of amenities provided with the listing, such as Wi-Fi or a kitchen.
# price: The nightly price for the listing.
# security_deposit: The amount of the security deposit, if required.
# cleaning_fee: The fee charged for cleaning the listing after a guest’s stay.
# extra_people: The fee for additional guests beyond the number included in the booking.
# guests_included: The number of guests included in the booking price.
# images: Photos of the listing.
# host: Information about the host, including their ID and profile details.
# address: The address of the listing.
# availability: The availability of the listing for booking.
# review_scores: The review scores for various aspects of the listing, such as cleanliness and communication.
# reviews: The text of the reviews left by guests.

**Data Extraction from Mongodb atlas**

In [6]:
data = {"Id" : [], "Name" : [], "Description" : [], "Neighbourhood_overview" : [], "Property_type" : [], "Room_type" : [], "Bed_type" : [], "Min_nights" : [], 
        "Max_nights" : [], "Cancellation_policy" : [], "Accommodates" : [], "No_of_reviews" : [], "Amenities" : [], "Host_id" : [], "Host_name" : [],
        "Host_neighbourhood" : [], "Street" : [], "Suburban" : [], "Region" : [], "Country" : [], "Longitude" : [], "Latitude" : [], "Availability_30days" : [],
        "Availability_60days" : [], "Availability_90days" : [], "Availability_365days" : [], "Price" : [], "Ratings" : []}

for i in my_coll.find():
    data["Id"].append(i["_id"])
    data["Name"].append(i["name"])
    data["Description"].append(i["description"])
    data["Neighbourhood_overview"].append(i["neighborhood_overview"])
    data["Property_type"].append(i["property_type"])
    data["Room_type"].append(i["room_type"])
    data["Bed_type"].append(i["bed_type"])
    data["Min_nights"].append(i["minimum_nights"])
    data["Max_nights"].append(i["maximum_nights"])
    data["Cancellation_policy"].append(i["cancellation_policy"])
    data["Accommodates"].append(i["accommodates"])
    data["No_of_reviews"].append(i["number_of_reviews"])
    data["Amenities"].append(i["amenities"])
    data["Host_id"].append(i["host"]["host_id"])
    data["Host_name"].append(i["host"]["host_name"])
    data["Host_neighbourhood"].append(i["host"]["host_neighbourhood"])
    data["Street"].append(i["address"]["street"])
    data["Suburban"].append(i["address"]["suburb"])
    data["Region"].append(i["address"]["market"])
    data["Country"].append(i["address"]["country"])
    data["Longitude"].append(i["address"]["location"]["coordinates"][0])
    data["Latitude"].append(i["address"]["location"]["coordinates"][1])
    data["Availability_30days"].append(i["availability"]["availability_30"])
    data["Availability_60days"].append(i["availability"]["availability_60"])
    data["Availability_90days"].append(i["availability"]["availability_90"])
    data["Availability_365days"].append(i["availability"]["availability_365"])
    data["Price"].append(i["price"])
    data["Ratings"].append(i["review_scores"].get("review_scores_rating"))

**Converting the Extracted Data into Pandas DataFrame**

In [7]:
df = pd.DataFrame(data)
df.head()

Id                              Name  \
0  10006546           Ribeira Charming Duplex   
1  10009999      Horto flat with small garden   
2   1001265  Ocean View Waikiki Marina w/prkg   
3  10021707          Private Room in Bushwick   
4  10030955       Apt Linda Vista Lagoa - Rio   

                                         Description  \
0  Fantastic duplex apartment with three bedrooms...   
1  One bedroom + sofa-bed in quiet and bucolic ne...   
2  A short distance from Honolulu's billion dolla...   
3  Here exists a very cozy room for rent in a sha...   
4  Quarto com vista para a Lagoa Rodrigo de Freit...   

                              Neighbourhood_overview Property_type  \
0  In the neighborhood of the river, you can find...         House   
1  This charming ground floor flat is located in ...     Apartment   
2             You can breath ocean as well as aloha.   Condominium   
3                                                        Apartment   
4                                                        Apartment   

         Room_type  Bed_type Min_nights Max_nights  \
0  Entire home/apt  Real Bed          2         30   
1  Entire home/apt  Real Bed          2       1125   
2  Entire home/apt  Real Bed          3        365   
3     Private room  Real Bed         14       1125   
4     Private room  Real Bed          1       1125   

           Cancellation_policy  ...          Region        Country  \
0                     moderate  ...           Porto       Portugal   
1                     flexible  ...  Rio De Janeiro         Brazil   
2  strict_14_with_grace_period  ...            Oahu  United States   
3                     flexible  ...        New York  United States   
4                     flexible  ...  Rio De Janeiro         Brazil   

    Longitude   Latitude Availability_30days Availability_60days  \
0   -8.613080  41.141300                  28                  47   
1  -43.230750 -22.966254                   0                   0   
2 -157.839190  21.286340                  16                  46   
3  -73.936150  40.697910                   0                   0   
4  -43.205047 -22.971951                  28                  58   

  Availability_90days Availability_365days   Price Ratings  
0                  74                  239   80.00    89.0  
1                   0                    0  317.00     NaN  
2                  76                  343  115.00    84.0  
3                   0                    0   40.00   100.0  
4                  88                  363  701.00     NaN  

[5 rows x 28 columns]

**Data Preprocessing**

In [8]:
df.shape

(5555, 28)

**The data has 5555 rows and 28 columns and in the above dataframe, we can see that some empty rows in some columns**

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5555 entries, 0 to 5554
Data columns (total 28 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Id                      5555 non-null   object 
 1   Name                    5555 non-null   object 
 2   Description             5555 non-null   object 
 3   Neighbourhood_overview  5555 non-null   object 
 4   Property_type           5555 non-null   object 
 5   Room_type               5555 non-null   object 
 6   Bed_type                5555 non-null   object 
 7   Min_nights              5555 non-null   object 
 8   Max_nights              5555 non-null   object 
 9   Cancellation_policy     5555 non-null   object 
 10  Accommodates            5555 non-null   int64  
 11  No_of_reviews           5555 non-null   int64  
 12  Amenities               5555 non-null   object 
 13  Host_id                 5555 non-null   object 
 14  Host_name               5555 non-null   

**Some Statistical Measure like Mean, Count,... for Numerical Columns**

In [10]:
df.describe()

Accommodates  No_of_reviews    Longitude     Latitude  \
count   5555.000000    5555.000000  5555.000000  5555.000000   
mean       3.505851      27.606481    -6.623240    22.061230   
std        2.297019      49.798376    90.525974    28.092112   
min        1.000000       0.000000  -159.678690   -34.088290   
25%        2.000000       1.000000   -73.599530    20.724390   
50%        3.000000       5.000000    -8.611420    40.727130   
75%        4.000000      32.000000    29.019605    41.162755   
max       16.000000     533.000000   151.338980    45.665630   

       Availability_30days  Availability_60days  Availability_90days  \
count          5555.000000          5555.000000          5555.000000   
mean             11.816202            26.451305            42.758056   
std              11.686113            23.476011            35.226897   
min               0.000000             0.000000             0.000000   
25%               0.000000             0.000000             0.000000   
50%               8.000000            23.000000            43.000000   
75%              24.000000            52.000000            80.000000   
max              30.000000            60.000000            90.000000   

       Availability_365days      Ratings  
count           5555.000000  4081.000000  
mean             173.105671    93.099240  
std              139.841893     9.023483  
min                0.000000    20.000000  
25%               17.000000    90.000000  
50%              171.000000    95.000000  
75%              317.000000    99.000000  
max              365.000000   100.000000

In [11]:
df.columns

Index(['Id', 'Name', 'Description', 'Neighbourhood_overview', 'Property_type',
       'Room_type', 'Bed_type', 'Min_nights', 'Max_nights',
       'Cancellation_policy', 'Accommodates', 'No_of_reviews', 'Amenities',
       'Host_id', 'Host_name', 'Host_neighbourhood', 'Street', 'Suburban',
       'Region', 'Country', 'Longitude', 'Latitude', 'Availability_30days',
       'Availability_60days', 'Availability_90days', 'Availability_365days',
       'Price', 'Ratings'],
      dtype='object')

 **To check if there is any duplicate rows in the data by using the id column as it will be a unique column in the data**

In [12]:
df[df["Id"].duplicated() == True]

Empty DataFrame
Columns: [Id, Name, Description, Neighbourhood_overview, Property_type, Room_type, Bed_type, Min_nights, Max_nights, Cancellation_policy, Accommodates, No_of_reviews, Amenities, Host_id, Host_name, Host_neighbourhood, Street, Suburban, Region, Country, Longitude, Latitude, Availability_30days, Availability_60days, Availability_90days, Availability_365days, Price, Ratings]
Index: []

[0 rows x 28 columns]

**There are no duplicate rows in the data**

In [13]:
df["Id"].nunique()

5555

**Below are the unique item for the Categorical features**

In [14]:
df["Property_type"].unique()

array(['House', 'Apartment', 'Condominium', 'Loft', 'Guesthouse',
       'Hostel', 'Serviced apartment', 'Bed and breakfast', 'Treehouse',
       'Bungalow', 'Guest suite', 'Townhouse', 'Villa', 'Cabin', 'Other',
       'Chalet', 'Farm stay', 'Boutique hotel', 'Boat', 'Cottage',
       'Earth house', 'Aparthotel', 'Resort', 'Tiny house',
       'Nature lodge', 'Hotel', 'Casa particular (Cuba)', 'Barn', 'Hut',
       'Camper/RV', 'Heritage hotel (India)', 'Pension (South Korea)',
       'Campsite', 'Houseboat', 'Castle', 'Train'], dtype=object)

In [15]:
df["Room_type"].unique()

array(['Entire home/apt', 'Private room', 'Shared room'], dtype=object)

In [16]:
df["Bed_type"].unique()

array(['Real Bed', 'Futon', 'Pull-out Sofa', 'Airbed', 'Couch'],
      dtype=object)

In [17]:
df["Cancellation_policy"].unique()

array(['moderate', 'flexible', 'strict_14_with_grace_period',
       'super_strict_30', 'super_strict_60'], dtype=object)

In [18]:
df["Country"].unique()

array(['Portugal', 'Brazil', 'United States', 'Turkey', 'Canada',
       'Hong Kong', 'Spain', 'Australia', 'China'], dtype=object)

In [19]:
df["Accommodates"].unique()

array([ 8,  4,  2,  1,  6,  3,  5,  9, 14, 11, 10,  7, 12, 16, 13, 15],
      dtype=int64)

In [20]:
df["Min_nights"].unique()

array(['2', '3', '14', '1', '12', '4', '15', '10', '7', '5', '21', '30',
       '28', '200', '32', '60', '31', '179', '6', '8', '180', '35', '13',
       '20', '160', '16', '90', '27', '45', '40', '29', '9', '96', '26',
       '25', '11', '365', '115', '100', '22', '300', '120', '80', '1250',
       '50'], dtype=object)

In [21]:
df["Max_nights"].unique()

array(['30', '1125', '365', '360', '75', '20', '9', '500', '21', '180',
       '7', '10', '6', '120', '15', '1124', '90', '300', '60', '3', '40',
       '89', '100', '28', '25', '45', '1120', '2147483647', '29', '32',
       '1123', '330', '730', '50', '179', '2', '1100', '5', '270', '31',
       '150', '1825', '14', '356', '600', '334', '92', '4', '23', '18',
       '59', '80', '8', '450', '33', '36', '5000', '22', '61', '24',
       '200', '1000', '19', '13', '99', '350', '190', '26', '12', '355',
       '160', '364', '999', '16', '1117', '900', '27', '1095', '114',
       '333333', '1109', '1126', '35', '700', '112', '181', '1234567890',
       '37', '56', '62', '548', '65', '366', '1001', '17', '1', '10000',
       '371', '1111', '11', '1003', '64', '720', '184', '140', '70', '91',
       '34', '103', '760', '69', '1110', '41', '250', '93', '1121', '185',
       '39', '130', '225', '85', '352', '240', '170', '79', '1084', '182',
       '1086', '1250', '55', '395', '2800', '305', '6

In [22]:
df["Street"].nunique()

677

In [23]:
df["Suburban"].nunique()

410

In [24]:
df["Region"].unique()

array(['Porto', 'Rio De Janeiro', 'Oahu', 'New York', 'Istanbul',
       'Montreal', 'Hong Kong', 'Barcelona', 'Sydney', 'The Big Island',
       'Maui', 'Kauai', '', 'Other (Domestic)', 'Other (International)'],
      dtype=object)

**Checking if there are any missing values in the columns**

In [25]:
df.isnull().sum()

Id                           0
Name                         0
Description                  0
Neighbourhood_overview       0
Property_type                0
Room_type                    0
Bed_type                     0
Min_nights                   0
Max_nights                   0
Cancellation_policy          0
Accommodates                 0
No_of_reviews                0
Amenities                    0
Host_id                      0
Host_name                    0
Host_neighbourhood           0
Street                       0
Suburban                     0
Region                       0
Country                      0
Longitude                    0
Latitude                     0
Availability_30days          0
Availability_60days          0
Availability_90days          0
Availability_365days         0
Price                        0
Ratings                   1474
dtype: int64

**we can see that Ratings feature has 1474 missing values. so, we will replace missing values with 0(zero)**

In [26]:
df["Ratings"] = df["Ratings"].fillna(value = 0)

In [27]:
df.isnull().sum()

Id                        0
Name                      0
Description               0
Neighbourhood_overview    0
Property_type             0
Room_type                 0
Bed_type                  0
Min_nights                0
Max_nights                0
Cancellation_policy       0
Accommodates              0
No_of_reviews             0
Amenities                 0
Host_id                   0
Host_name                 0
Host_neighbourhood        0
Street                    0
Suburban                  0
Region                    0
Country                   0
Longitude                 0
Latitude                  0
Availability_30days       0
Availability_60days       0
Availability_90days       0
Availability_365days      0
Price                     0
Ratings                   0
dtype: int64

**Checking if there are any empty strings in any of the columns**

In [28]:
for i in  df.columns:
    print(i,"--->",(df[i] == "").sum())

Id ---> 0
Name ---> 8
Description ---> 95
Neighbourhood_overview ---> 2241
Property_type ---> 0
Room_type ---> 0
Bed_type ---> 0
Min_nights ---> 0
Max_nights ---> 0
Cancellation_policy ---> 0
Accommodates ---> 0
No_of_reviews ---> 0
Amenities ---> 0
Host_id ---> 0
Host_name ---> 0
Host_neighbourhood ---> 1923
Street ---> 0
Suburban ---> 887
Region ---> 6
Country ---> 0
Longitude ---> 0
Latitude ---> 0
Availability_30days ---> 0
Availability_60days ---> 0
Availability_90days ---> 0
Availability_365days ---> 0
Price ---> 0
Ratings ---> 0


**There are six features which has a empty string in their columns. so, we will try to replace the empty strings by a placeholder("Unknown") for all other features without including Region feature**

**For Region feature, we will get the Street column of that empty string row and will try to match the Street to get the unique value of region column and replace the values with the unique value other than Empty string.** 

In [29]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "Waverley, NSW, Australia"), "Region"] = df.loc[df["Street"] == "Waverley, NSW, Australia", "Region"].unique()[1]  # 'Sydney'

In [30]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "New York, NY, United States"), "Region"] = df.loc[df["Street"] == "New York, NY, United States", "Region"].unique()[0]  # 'New York'

In [31]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "Rio de Janeiro, Rio de Janeiro, Brazil"), "Region"] = df.loc[df["Street"] == "Rio de Janeiro, Rio de Janeiro, Brazil", "Region"].unique()[0]  # 'Rio De Janeiro'

In [32]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "Montréal, Québec, Canada"), "Region"] = df.loc[df["Street"] == "Montréal, Québec, Canada", "Region"].unique()[0]   # 'Montreal'

In [33]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "Barcelona, Catalunya, Spain"), "Region"] = df.loc[df["Street"] == "Barcelona, Catalunya, Spain", "Region"].unique()[0]  # 'Barcelona'

In [34]:
df.loc[(df["Region"] == "" ) & (df["Street"] == "Kapaa, HI, United States"), "Region"] = df.loc[df["Street"] == "Kapaa, HI, United States", "Region"].unique()[0]  # 'Kauai'

In [35]:
df.loc[df["Name"] == "", "Name"] = "Unknown"

In [36]:
df.loc[df["Description"] == "", "Description"] = "Unknown"

In [37]:
df.loc[df["Neighbourhood_overview"] == "", "Neighbourhood_overview"] = "Unknown"

In [38]:
df.loc[df["Host_neighbourhood"] == "", "Host_neighbourhood"] = "Unknown"

In [39]:
df.loc[df["Suburban"] == "", "Suburban"] = "Unknown"

Now, there are no empty strings in the dataset

In [40]:
for i in  df.columns:
    print(i,"--->",(df[i] == "").sum())

Id ---> 0
Name ---> 0
Description ---> 0
Neighbourhood_overview ---> 0
Property_type ---> 0
Room_type ---> 0
Bed_type ---> 0
Min_nights ---> 0
Max_nights ---> 0
Cancellation_policy ---> 0
Accommodates ---> 0
No_of_reviews ---> 0
Amenities ---> 0
Host_id ---> 0
Host_name ---> 0
Host_neighbourhood ---> 0
Street ---> 0
Suburban ---> 0
Region ---> 0
Country ---> 0
Longitude ---> 0
Latitude ---> 0
Availability_30days ---> 0
Availability_60days ---> 0
Availability_90days ---> 0
Availability_365days ---> 0
Price ---> 0
Ratings ---> 0


**checking if the column types are in correct format that will be useful for analysis**

In [41]:
df.dtypes

Id                         object
Name                       object
Description                object
Neighbourhood_overview     object
Property_type              object
Room_type                  object
Bed_type                   object
Min_nights                 object
Max_nights                 object
Cancellation_policy        object
Accommodates                int64
No_of_reviews               int64
Amenities                  object
Host_id                    object
Host_name                  object
Host_neighbourhood         object
Street                     object
Suburban                   object
Region                     object
Country                    object
Longitude                 float64
Latitude                  float64
Availability_30days         int64
Availability_60days         int64
Availability_90days         int64
Availability_365days        int64
Price                      object
Ratings                   float64
dtype: object

**Columns like Min_nights, Max_nights and Price are not in correct format. we will change the  data types as below**

In [42]:
df["Min_nights"] = df["Min_nights"].astype(int)

In [43]:
df["Max_nights"] = df["Max_nights"].astype(int)

In [44]:
df["Price"] = df["Price"].astype(str).astype(float)

**Finally, As all the preprocessing steps are done , now will save the data to csv format**

In [45]:
df.to_csv("Airbnb Dataset.csv", index = False, encoding = "utf-8-sig")